Chapter controls for the seek bar

jonasburian opened the pull request:
This adds the possibility to change the current chapter by tapping on the corresponding area on the seek bar as proposed in #4898. In addition, the current chapter title will be shown in the popup for the current playback time while seeking.

ByteHamster commented:
Nice, thanks for working on this. I have not looked at the code yet but I tried it on my device. Some comments:

- Nice idea that double-tapping switches to a position even if it is in the middle of a chapter!
- The popup only shows the chapter title. I think it would be better to display both position and chapter title. If the position is displayed below the title, it will not jump around when the title length changes. 
- Especially for long chapters, I think it feels a bit strange that the seekbar thumb does not go where you touched it. How about highlighting the pressed section for a short time (like it is done when seeking)? Not sure how easy it would be to implement.

jonasburian commented:
Thanks for your feedback! 
> double-tapping switches to a position

The actual behavior is, that within a chapter you can switch to every position within the chapter. So it would switch to the beginning of the chapter first and than to the desired position :D.

>  it will not jump around when the title length changes

I think the problem is, that a text that goes over two lines takes the full width of cardView. If I just add a `\n` it won't work. I think it is possible to change the width of the cardView to `match_parent` programmatically, but it worked not as expected for now. I will look into it a bit more.

> highlighting the pressed section for a short time

There seems to be a problem with the positioning of the chapter start positions when I use `seekToChapter`, so that sometimes the previous chapter is highlighted. Interestingly, the positioning of the dividers on the seek bar and the progress within the chaptersFragment seems to be even.

ByteHamster commented:
> I think the problem is, that a text that goes over two lines takes the full width of the RelativeLayout. If I just add a \n it won't work.

Why is it a problem? The text is centered, so shouldn't it stay in the same position? Maybe I'm misunderstanding something.

> There seems to be a problem with the positioning of the chapter start positions when I use seekToChapter, so that sometimes the previous chapter is highlighted. 

I changed the color of the buffering indicator to be the same as the chapter highlighting (because the indicator was extremely strong). Could that be the reason that the previous chapter looks like it is highlighted?

jonasburian commented:
> Maybe I'm misunderstanding something.

No you're right. I misunderstood something :). Fixed it with the last commit

> Could that be the reason that the previous chapter looks like it is highlighted?

I think I found the reason. It's probably the `updatePosition` method (rounding errors caused by it). I think I have to call 
`seekToChapter` in another way. I'll work on this in the next couple of days.

> I changed the color of the buffering indicator

Thanks btw. :)

jonasburian commented:
I'm a bit out of ideas with why it behaves like it does :/. So it seems like it seeks to the right position, but then updates the position again and it is no longer accurate.

ByteHamster commented:
How far does it rewind? Could this be related to the auto-rewind feature when continuing playback after a long time? If that's the case, the `develop` branch without your PR should have similar jumps.

jonasburian commented:
I think it is more position related, something of a correction of the position by the seek bar. It does not seek/rewind very far. At max I would say 30 sec, but it increases to the beginning and to the end. The problem is, that it has to be at the exact position to highlight the correct chapter.

ByteHamster commented:
Sorry for the late reply!

I think I know what the problem is. In `onProgressChanged`, you seek to the chapter. Then, afterwards, the SeekBar calls `onStopTrackingTouch`, which also seeks. The behavior of the two seek operations is slightly different, though. Your chapter seeking code uses the exact timestamps that are specified by the publishers. The seeking code in `onStopTrackingTouch` uses only an estimate (0.2% accuracy because the SeekBar has 500 steps at most. Also, the SeekBar only uses an estimate for the total duration, I think, which might not be 100% accurate when using variable bit rate encoding).

I think it would be better to seek to the chapter in `onStopTrackingTouch` instead, but only if the total distance sought (or tap duration?) is very small and therefore represents a single click. That way, only one single seek operation is performed, which should make it more consistent.

jonasburian commented:
Thanks for your help! It works now :). (I took a slightly different approach to make it more responsive)

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/AudioPlayerFragment.java:
```diff
@@ -495,7 +501,28 @@ public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
+                        public void run() {
+                            sbPosition.setIsEnabled(false);
+                        }
+                    }, 5000);
```
- When clicking multiple chapters directly after each other, the handlers could interfere. I think it would be better to move this to a `sbPosition.highlightCurrentChapter()` method that posts the runnable internally. It could then keep a fixed `Handler` instance and cancel other scheduled tasks before posting a new one.
- I find `setIsEnabled` a bit misleading because it reads like the framework method `setEnabled`. Will be fixed automatically by the comment above, though :)
- I have not tried running it on my device but I think 5 seconds might be a bit too much. How about 1 second?

jonasburian replied:
Thanks! The last commit should fix it. I had to call `invalidate()` so it is really only highlighted for a second. (And one second looks good to me :D)

ByteHamster reviewed (commented):
Thanks! Another round of comments :)

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/view/ChapterSeekBar.java:
```diff
@@ -59,6 +62,17 @@ public void setDividerPos(final float[] dividerPos) {
     }
 
+    public void highlightCurrentChapter() {
+        isEnabled = true;
```
Could you please rename the variable? I think it is a bit misleading. I would suggest `isHighlighted`.

jonasburian replied:
Fixed :)

ByteHamster commented:
Awesome, thanks! Will be released in 2.3.0 :)

By the way, I noticed a problem with the seek bar while testing another PR. I thought I already commented that here but apparently I forgot to actually press the "comment" button. The seek bar looks a bit broken on an API 19 emulator, see image below. Could you please have a look?

![image](https://user-images.githubusercontent.com/5811634/114783671-5aceb800-9d7a-11eb-83a9-8b2e85792760.png)

In order to test on API 19, you need to replace `extends Application` with `extends MultiDexApplication` in `PodcastApp.java`.

jonasburian commented:
Thanks for that! Took a quick look into it. From what I've been able to determine so far, the behavior when changing the parameters for the height looks a bit weird between the two API versions, because it seems to be the contrary. I need to look into it a bit more within the next couple of days
